### PR TITLE
Fix MIDI note-on events being converted to note-off events

### PR DESCRIPTION
**Pull request by voidshine**
This PR fixes a bug in Godot's MIDI processing logic that destroys information as it flows from device to application. The engine converts note-on events with velocity 0 to note-off for programmer convenience, which is sometimes helpful and sometimes confounding. In fact, MIDI device implementations vary and some MIDI devices intend for note-on to mean note-on regardless of velocity. So the only way to support such devices is to faithfully pass MIDI information to the application exactly as it is received from the device.

It's up to application developers to decide how to handle received MIDI data and it's a very minor inconvenience to convert if needed, but having the engine interfere and mutate event data before it reaches the application makes some use cases impossible.

I commented on this before but the issues were closed, so now I'm submitting a fix directly:
* https://github.com/godotengine/godot/issues/27199#issuecomment-942071469
* https://github.com/godotengine/godot/pull/27214#issuecomment-942063417

### CODE REVIEW

**Comment by akien-mga** [comment]
The documentation should likely be improve to advise users that they have to do this check themselves if they rely on this common convention.

**Comment by voidshine** [comment]
Good idea, thanks -- I added a second commit that updates the documentation.

**Comment by reduz** [comment]
AFAIK I am not sure what the point of this is, NOTE ON with velocity 0 is always expected to be note off. Any synth will react to this.

**Comment by reduz** [comment]
@voidshine makes sense, I guess if you want to use a device as a controller for a specific situation, Godot should simply pass what comes in, as there is no reason to alter it.

**Review by reduz — APPROVED** [review]

File: doc/classes/InputEventMIDI.xml
Comment (akien-mga) [inline doc/classes/InputEventMIDI.xml:93]:
```suggestion
			The velocity of the MIDI signal. This value ranges from 0 to 127. For a piano, this corresponds to how quickly the key was pressed, and is rarely above about 110 in practice. Note that some MIDI devices may send a [constant MIDI_MESSAGE_NOTE_ON] message with zero velocity and expect this to be treated the same as a [constant MIDI_MESSAGE_NOTE_OFF] message, but device implementations vary so Godot reports event data exactly as received.
```
Please do the change by amending the commit with `git commit --amend` (and then `git push --force`) to keep it as a single commit ready to merge.

**Comment by akien-mga** [comment]
For the record, the email you used to author this commit isn't valid, so it's not attributed to your GitHub account (see the header of https://patch-diff.githubusercontent.com/raw/godotengine/godot/pull/66003.patch). If you want to change it, you can do so by editing your local Git identity and then `git commit --amend --reset-author`).

File: doc/classes/InputEventMIDI.xml
Comment (voidshine) [reply doc/classes/InputEventMIDI.xml]: I made these suggested changes and force pushed. Also corrected the email address. Thank you!

**Comment by akien-mga** [comment]
Could you squash the 2 commits? See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) for instructions.

(For the record that's why I suggested *amending* the original commit with `git commit --amend`, instead of making a new commit.)

That would also give you the possibility to fix the email on that first commit, which is still the original invalid one.

And finally, the new email you used is valid, but not tied to your GitHub account. That's not a problem, but if you want the commit to appear as authored by @voidshine, you can add that email as secondary email in your GitHub settings.

**Comment by voidshine** [comment]
@akien-mga I actually did use `--amend`, I entered the git commands as written. So there should only be my original two commits, no extras. Are you saying I should not have created the second documentation-update commit in the first place? Having two seemed sensible at the time and doesn't seem to violate the article's principle about keeping Godot in a functional state. Happy to conform to Godot's PR workflow here, I just thought what I did was sensible.

Would squashing the code and documentation commits together invalidate the approval? I'd like to avoid spending more code approver time on a history change if possible. If that's allowed without approval invalidation, I can try my hand at rewriting history...

**Comment by voidshine** [comment]
Also, thanks again for the tip about email -- I am not too concerned about credit here, but will go ahead and validate the email with GitHub so it will link things up.

**Comment by akien-mga** [comment]
Ah I see. Well the code change and the documentation are related, so it's better to have them in a single commit. The documentation change is only valid because of the code change so they can be kept together as a unit.

**Comment by voidshine** [comment]
Makes sense, I just did the squash, thanks for the guidance here!

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by voidshine** [comment]
Thanks! :tada: Hoping to make some more in years to come. :heart_decoration: Godot!

**Comment by akien-mga** [comment]
Cherry-picked for 3.6.

**Pull request merged**
